vllm/sequence.py

```diff
@@ -52,6 +53,9 @@ class SequenceData:
         prompt_token_ids: The token IDs of the prompt.
         output_token_ids: The token IDs of the output.
         cumulative_logprob: The cumulative log probability of the output.
+        prompt_logprobs: The log probabilities of the prompt tokens.
+        prompt_top_logprobs: The log probabilities of the top probability
+        tokens in the prompt.
```

Comment:
Can we have only one variable `prompt_logprobs`? similar to `output_logprobs` in the `Sequence` class.

Reply:
In OpenAI implementation, they allow `logprobs=0` while getting back the only logprobs chosen for the prompt. If we only keep one variable, we won't be able to get the logprob for the chosen tokens since they might not be in our top-k.

Reply:
We support the OpenAI's behavior and [this is what we did](https://github.com/vllm-project/vllm/blob/d6545ad22ee89c8b6e3eb6cfcf8ff914a06ccee1/vllm/model_executor/layers/sampler.py#L383-L393) for `output_logprobs`: We will select the top-k (can be top-0), and also force the generated token to be in the `output_logprobs` dict. I believe we can do the same thing for `prompt_logprobs`.

Reply:
But this might result in different sizes of `logprobs`? Say if we want `logprobs=2`, then for some tokens, it might return 3? or we want another filter at the end?

Reply:
> if we want logprobs=2, then for some tokens, it might return 3

Yes, and this is OpenAI's behavior as well.

Reply:
I just verified, it isn't. 

Request:
```
{
    "model": "text-davinci-003",
    "prompt": "Please get what I booked",
    "echo": true,
    "n": 1,
    "temperature": 1.0,
    "max_tokens": 1,
    "logprobs": 2
}
```

Response:
```
{
    "object": "text_completion",
    "created": 1694560209,
    "model": "text-davinci-003",
    "choices": [
        {
            "text": "Please get what I booked\n",
            "index": 0,
            "logprobs": {
                "tokens": [
                    "Please",
                    " get",
                    " what",
                    " I",
                    " booked",
                    "\n"
                ],
                "token_logprobs": [
                    null,
                    -6.516949,
                    -7.269826,
                    -1.5403796,
                    -11.766179,
                    -0.064088084
                ],
                "top_logprobs": [
                    null,
                    {
                        " enter": -1.8900735,
                        " note": -2.4450452
                    },
                    {
                        " in": -1.1974856,
                        " your": -2.421246
                    },
                    {
                        " I": -1.5403796,
                        " you": -0.7000267
                    },
                    {
                        " need": -1.2898092,
                        "'m": -1.5458908
                    },
                    {
                        "\n": -0.064088084,
                        " for": -3.8346472
                    }
                ],
                "text_offset": [
                    0,
                    6,
                    10,
                    15,
                    17,
                    24
                ]
            },
            "finish_reason": "length"
        }
    ],
    "usage": {
        "prompt_tokens": 5,
        "completion_tokens": 1,
        "total_tokens": 6
    }
}
```

The prompt tokens I used are not in the returned `top_logprobs` while they are fixed at the size of 2.

Reply:
I see. We implemented vLLM based on [the description in the API doc](https://platform.openai.com/docs/api-reference/completions/create#logprobs).

However, I believe two variables introduce unnecessary complexity. Let's just use one variable in the code. If we really want to be exactly the same as OpenAI's current behavior, let's maybe add a filter in the end as you suggested?